backend/infrahub/proposed_change/tasks.py

```diff
@@ -2,15 +2,109 @@
+#         proposed_change_id=flow_run.parameters["proposed_change_id"], state=ProposedChangeState.MERGED
+#     )
+
+
+# async def proposed_change_transition_open(flow: Flow, flow_run: FlowRun, state: State) -> None:
+#     await _proposed_change_transition_state(
+#         proposed_change_id=flow_run.parameters["proposed_change_id"], state=ProposedChangeState.OPEN
+#     )
+
+
+@flow(
+    name="proposed-change-merge",
+    flow_run_name="Merge propose change: {proposed_change_name} ",
+    description="Merge a given proposed change.",
+    # TODO need to investigate why these function are not working as expected
+    # on_completion=[proposed_change_transition_merged],  # type: ignore
+    # on_failure=[proposed_change_transition_open],  # type: ignore
+    # on_crashed=[proposed_change_transition_open],  # type: ignore
+    # on_cancellation=[proposed_change_transition_open],  # type: ignore
+)
+async def merge_proposed_change(proposed_change_id: str, proposed_change_name: str) -> State:  # pylint: disable=unused-argument
+    service = services.service
+    log = get_run_logger()
+
+    async with service.database.start_session() as db:
+        proposed_change = await registry.manager.get_one(
+            db=db, id=proposed_change_id, kind=InternalCoreProposedChange, raise_on_error=True
+        )
+
+        log.info("Validating if all conditions are met to merge the proposed change")
+
+        source_branch = await Branch.get_by_name(db=db, name=proposed_change.source_branch.value)
+        validations = await proposed_change.validations.get_peers(db=db, peer_type=CoreValidator)
+        for validation in validations.values():
+            validator_kind = validation.get_kind()
+            if (
+                validator_kind != InfrahubKind.DATAVALIDATOR
+                and validation.conclusion.value.value != ValidatorConclusion.SUCCESS.value
+            ):
+                # Ignoring Data integrity checks as they are handled again later
+                await _proposed_change_transition_state(proposed_change=proposed_change, state=ProposedChangeState.OPEN)
+                return Failed(message="Unable to merge proposed change containing failing checks")
+            if validator_kind == InfrahubKind.DATAVALIDATOR:
+                data_checks = await validation.checks.get_peers(db=db, peer_type=CoreDataCheck)
+                for check in data_checks.values():
+                    if check.conflicts.value and not check.keep_branch.value:
+                        await _proposed_change_transition_state(
+                            proposed_change=proposed_change, state=ProposedChangeState.OPEN
+                        )
+                        return Failed(
+                            message="Data conflicts found on branch and missing decisions about what branch to keep"
+                        )
+
+        log.info("Proposed change is eligible to be merged")
+        await merge_branch(branch=source_branch.name)
```

Comment:
we can do both
using a workflow with the `submit_workflow` or `execute_workflow` will leverage Prefect to schedule the function to be executed on the next available worker
Using the function directly like I'm doing here will execute the code directly on the worker but the reporting will still go to prefect
from the user point of view it looks the same.
it's more a question of where we want the execution to be done
in this case since we need to wait for it to finish anyway, we might as well do it on the same worker and same the time it takes to schedule a workflow